controllers/common/condition/controller.go

```diff
@@ -139,5 +102,68 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		r.Recorder.Eventf(obj, "Normal", "Failed", "Failed to update conditions: %s", updateError.Error())
 		return ctrl.Result{}, nil
 	}
+
 	return ctrl.Result{}, nil
 }
+
+func diffConditions(obj v1alpha1.InnerObject) (newConditionMap map[v1alpha1.ChaosConditionType]StatusAndReason) {
+	records := obj.GetStatus().Experiment.Records
+	newConditionMap = make(map[v1alpha1.ChaosConditionType]StatusAndReason)
+
+	if records != nil {
+		newConditionMap[v1alpha1.ConditionSelected] = StatusAndReason{
+			Status: corev1.ConditionTrue,
+		}
+	} else {
+		newConditionMap[v1alpha1.ConditionSelected] = StatusAndReason{
+			Status: corev1.ConditionFalse,
+		}
+	}
+
+	// If records is `nil`, we don't need to check the `allInjected` and `allRecovered` conditions.
+	var allInjected corev1.ConditionStatus
+	if records != nil && every(records, func(record *v1alpha1.Record) bool {
+		return record.Phase == v1alpha1.Injected
+	}) {
+		allInjected = corev1.ConditionTrue
+	} else {
+		allInjected = corev1.ConditionFalse
+	}
+
+	var allRecovered corev1.ConditionStatus
+	if records != nil && every(records, func(record *v1alpha1.Record) bool {
+		return record.Phase == v1alpha1.NotInjected
+	}) {
+		allRecovered = corev1.ConditionTrue
+	} else {
+		allRecovered = corev1.ConditionFalse
+	}
+
+	newConditionMap[v1alpha1.ConditionAllInjected] = StatusAndReason{
+		Status: allInjected,
+	}
+	newConditionMap[v1alpha1.ConditionAllRecovered] = StatusAndReason{
+		Status: allRecovered,
+	}
+
+	if obj.IsPaused() {
+		newConditionMap[v1alpha1.ConditionPaused] = StatusAndReason{
+			Status: corev1.ConditionTrue,
+		}
+	} else {
+		newConditionMap[v1alpha1.ConditionPaused] = StatusAndReason{
+			Status: corev1.ConditionFalse,
+		}
+	}
+
+	return
+}
+
+func every[T any](arr []T, condition func(T) bool) bool {
```

Comment:
Can you rename this function? Let us know roughly what this function does from the name

Reply:
This function is generic and usually tests whether all elements in the slice pass the condition function, and it needs to be put into context to know what it is used for:

```go
every(records, func(record *v1alpha1.Record) bool {
        return record.Phase == v1alpha1.Injected
})
```

I don't think there's a better name for it, so I'll go ahead and add a comment to explain what it's for.

Reply:
[7c91f92](https://github.com/chaos-mesh/chaos-mesh/pull/4199/commits/7c91f928ba5f84de84140e5a8ad4b9bad5c18ddc)